core/elf.cc

```diff
@@ -791,9 +792,9 @@ void object::relocate_pltgot()
     pltgot[1] = this;
     pltgot[2] = reinterpret_cast<void*>(__elf_resolve_pltgot);
 
-    bool bind_now = dynamic_exists(DT_BIND_NOW) || mlocked() ||
+    bool bind_now = dynamic_exists(DT_BIND_NOW) ||
```

Comment:
The new `libvdso.so` is built with `BIND_NOW` so we can have symbols resolved eagerly. However, there are many old and new ways GCC marks it as you can see in this elaborate `if` expression which I think you expanded on at some point (see cc6e8f59425bc32bc2944856324b81b287d50390). And because of that the `mlocked()` would be evaluated second (before the remaining ones) and would try to call `sections()` which would eventually call `read()` which throws an exception for `memory_image`. So I could have fixed it differently, but changing the order was the easiest and possibly the right thing to do as `mlocked()` is really the OSv-specific "last-resort" type of scenario so it should be treated as such.